### PR TITLE
Add Regal for linting Rego

### LINKED ISSUE #559: Add in the OPA linter
# 💡 Summary #

What is the work, as a high-level summary?
The OPA team has recently released a version of a OPA linter that they are satisfied with. Now that is is stable & fairly roubust, should we include the linter in our CI/CD pipeline?

## Motivation and context ##

Why does this work belong in this project?
This linter allows for ignoring specific files & rules as well as only triggering warnings. Right now we have a style guide for the rego, but it is not enforced. With the linter we do not have to manually check for syntax errors, improving the readability of the code.

## Implementation notes ##

Please provide details for implementation, such as:

- https://github.com/StyraInc/regal

## Acceptance criteria ##

- [ ] Linter added into the pipeline
- [ ] Code updated for OPA style requierments

### PR DESCRIPTION
**Pull request by anderseknert**
## 🗣 Description ##

This PR adds [Regal](https://github.com/styraInc/regal) the Rego linter, to lint this project on PRs. Violations against Regal rules will be presented as annotations in PRs, and the full report printed to the logs.

I took the liberty of fixing some hopefully uncontroversial rules:

* [use-assignment-operator](https://docs.styra.com/regal/rules/style/use-assignment-operator)
* [use-in-operator](https://docs.styra.com/regal/rules/idiomatic/use-in-operator)
* [unconditional-assignment](https://docs.styra.com/regal/rules/style/unconditional-assignment)
* [file-missing-test-suffix](https://docs.styra.com/regal/rules/testing/file-missing-test-suffix)

The rest of the violations reported have been configured to level `ignore` for now, and I've annotated the configuration file here with some explanations for most of them, should you want to look into having more of them enforced.

Finally, I've updated the OPA test job run in CI to include linting with Regal.

Fixes #559

## ✅ Pre-approval checklist ##

- [x] This PR has an informative and human-readable title.
- [x] Changes are limited to a single goal - *eschew scope creep!*
- [x] *All* future TODOs are captured in issues, which are referenced
      in code comments.
- [ ] All relevant type-of-change labels have been added.
- [x] I have read the [CONTRIBUTING](../blob/develop/CONTRIBUTING.md) document.
- [x] These code changes follow [cisagov code standards](https://github.com/cisagov/development-guide).
- [x] All relevant repo and/or project documentation has been updated
      to reflect the changes in this PR.
- [x] Tests have been added and/or modified to cover the changes in this PR.
- [x] All new and existing tests pass.

## ✅ Pre-merge checklist ##

- [ ] Revert dependencies to default branches.
- [ ] Finalize version.

## ✅ Post-merge checklist ##

- [ ] Create a release.

### CODE REVIEW

File: .regal/config.yaml
Comment (gdasher) [inline .regal/config.yaml:1]: If you wouldn't mind it, after this is committed, please file an issue for us to fix at least these style issues:

1) line length
2) identically named tests

**Review by gdasher — COMMENTED** [review]
Thanks for doing this!  Overall this looks good and the changes to the rego appear fairly minor.

I will approve for merge after at least one of the main devs working on the project takes a look (cassie or david).

File: .regal/config.yaml
Comment (anderseknert) [reply .regal/config.yaml:1]: Will do! 👍 

I'd suggest adding the [constant condition](https://docs.styra.com/regal/rules/bugs/constant-condition) rule to that list too, but that's obviously not my call.
The way this project sometimes use

```rego
some_rule[message] {
    true
}
```

would better be expressed as

```rego
some_rule[message]
```

Or using more recent syntax

```
some_rule contains message
```

Checking the AST of these constructs (`opa parse --format json policy.rego`) shows these are all identical construct.

Having that `true` in the rule body is harmless, _but_ there are other places where a constant condition most certainly is a mistake, and having that rule disabled won't catch those. I guess they might be easy enough to spot in reviews though.

**Comment by anderseknert** [comment]
Thanks @gdasher! I've learnt quite a few things in the process, and I've created a couple of issue for suggesting new rules to Regal from this. So it's a win-win really :) I'd be happy to rebase later when the larger changes have made it in. Just say when 👍

File: .regal/config.yaml
Comment (anderseknert) [reply .regal/config.yaml:1]: Actually found an [issue](https://github.com/StyraInc/regal/issues/399) in Regal while looking into this, so let's leave these rules as they are until we can bump the version to the next one including the [fix](https://github.com/StyraInc/regal/pull/400) 😅

**Comment by gdasher** [comment]
Great.  Just waiting for #584 and #602 to land before approving this--we're trying to hit a scheduled release and so bandwidth is tight.  You may want to rebase to current HEAD on emerald and resolve the conflicts from a few recent commits in the meantime though.

**Comment by anderseknert** [comment]
@gdasher done! I've also bumped Regal to v0.11.0 released today, and made a few updates from the new rules there. Let me know when I should rebase again 👍

**Comment by gdasher** [comment]
Thanks.  Can you rebase one last time?  We had one final PR go in that added a bit more rego code in defender.  After that, we should be good to merge this.

Thanks for bearing with us--you submitted this at a slightly awkward time, but I appreciate your perseverence.

**Review by gdasher — APPROVED** [review]
LGTM

@Sloane4 Please take a look if you want to.  I will discuss with others on when to merge/how this given the other goings on.

**Comment by gdasher** [comment]
FYI, the smoke test failure is because it depends on a github secret that doesn't work from the forked repository being used in this commit [1].  But the other tests should give us needed confidence and the smoketest should run successfully on the commit itself.


[1] https://docs.github.com/en/actions/security-guides/using-secrets-in-github-actions

**Comment by anderseknert** [comment]
Thanks @gdasher 👍 I'm happy to help should there be any follow-up questions. Either ping me here, or there's a few ways to reach me listed in my GitHub profile. There's also a #regal channel in our [community Slack](https://communityinviter.com/apps/styracommunity/signup), should anyone want to get more involved in the project, get some help writing custom rules, or whatnot.

**Comment by anderseknert** [comment]
@buidav rebased and pushed 👍

**Comment by anderseknert** [comment]
Got notified about a failing test in the **OPA Tests** workflow in my fork. That seems to be failing on the `emerald` branch I rebased on too though, so I don't think it's related to any change here.

Screenshot 2023-10-25 at 23 07 30

**Pull request merged**
